[clang-tidy] Segmentation fault during check modernize-loop-convert

I get a Segmentation fault when analyzing the following code with clang-tidy 17.0.2 with the check `modernize-loop-convert` in about 50% of the runs:

test.cpp:
```c++
template <unsigned int p> void test() {
  unsigned int test[3][3 * p];

  // Initialize to zero
  for (unsigned int i = 0; i < 3 * p; ++i)
    for (unsigned int j = 0; j < 3; ++j)
      test[j][i] = 0;
}
```

by running `clang-tidy test.cpp -checks=modernize-loop-convert`.

In around 50% of the cases, I get the following error:
```
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /home/gebauer/programs/clang-tidy-17.0.2/usr/local/bin/clang-tidy test.cpp -checks=modernize-loop-convert
1.      <eof> parser at end of file
2.      ASTMatcher: Processing 'modernize-loop-convert' against:
        ForStmt : </home/gebauer/work/baci/test/test.cpp:6:5, line:7:20>
--- Bound Nodes Begin ---
    conditionBound - { ImplicitCastExpr : </home/gebauer/work/baci/test/test.cpp:6:34> }
    forLoopArray - { ForStmt : </home/gebauer/work/baci/test/test.cpp:6:5, line:7:20> }
    initVar - { VarDecl j : </home/gebauer/work/baci/test/test.cpp:6:10, col:27> }
--- Bound Nodes End ---
  #0 0x00005590c08d407f llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/gebauer/programs/clang-tidy-17.0.2/usr/local/bin/clang-tidy+0x453207f)
  #1 0x00005590c08d1af4 SignalHandler(int) Signals.cpp:0:0
  #2 0x00007f36de74d420 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14420)
  #3 0x00005590bf6deb5a clang::ASTContext::getTypeInfo(clang::Type const*) const (/home/gebauer/programs/clang-tidy-17.0.2/usr/local/bin/clang-tidy+0x333cb5a)
  #4 0x00005590bf6f270f clang::ASTContext::getTypeInfoImpl(clang::Type const*) const (/home/gebauer/programs/clang-tidy-17.0.2/usr/local/bin/clang-tidy+0x335070f)
  #5 0x00005590bf6deca5 clang::ASTContext::getTypeInfo(clang::Type const*) const (/home/gebauer/programs/clang-tidy-17.0.2/usr/local/bin/clang-tidy+0x333cca5)
[...]
#254 0x00005590bf6f270f clang::ASTContext::getTypeInfoImpl(clang::Type const*) const (/home/gebauer/programs/clang-tidy-17.0.2/usr/local/bin/clang-tidy+0x335070f)
#255 0x00005590bf6deca5 clang::ASTContext::getTypeInfo(clang::Type const*) const (/home/gebauer/programs/clang-tidy-17.0.2/usr/local/bin/clang-tidy+0x333cca5)
Segmentation fault (core dumped)
```

For the other 50% of the runs, I get the expected result:
```
1 warning generated.
/home/gebauer/work/baci/test/test.cpp:6:5: warning: use range-based for loop instead [modernize-loop-convert]
    6 |     for (unsigned int j = 0; j < 3; ++j)
      |     ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |         (auto & j : test)
    7 |       test[j][i] = 0;
      |       ~~~~~~~
      |       j
```

I'm running this on Ubuntu 20.04.

--- a/clang-tools-extra/clang-tidy/modernize/LoopConvertCheck.cpp
+++ b/clang-tools-extra/clang-tidy/modernize/LoopConvertCheck.cpp
@@ -753,6 +753,7 @@ void LoopConvertCheck::doConversion(
   bool IsCheapToCopy =
       !Descriptor.ElemType.isNull() &&
       Descriptor.ElemType.isTriviallyCopyableType(*Context) &&
+      !Descriptor.ElemType->isDependentSizedArrayType() &&
       // TypeInfo::Width is in bits.
       Context->getTypeInfo(Descriptor.ElemType).Width <= 8 * MaxCopySize;
   bool UseCopy = CanCopy && ((VarNameFromAlias && !AliasVarIsRef) ||

--- a/clang-tools-extra/docs/ReleaseNotes.rst
+++ b/clang-tools-extra/docs/ReleaseNotes.rst
@@ -267,11 +267,8 @@ Changes in existing checks
 
 - Improved :doc:`modernize-loop-convert
   <clang-tidy/checks/modernize/loop-convert>` to support for-loops with
-  iterators initialized by free functions like ``begin``, ``end``, or ``size``.
-
-- Improved :doc:`modernize-loop-convert
-   <clang-tidy/checks/modernize/loop-convert>` to avoid crash dor dependent
-   array.
+  iterators initialized by free functions like ``begin``, ``end``, or ``size``
+  and avoid crash for array of dependent array.
 
 - Improved :doc:`modernize-return-braced-init-list
   <clang-tidy/checks/modernize/return-braced-init-list>` check to ignore

--- a/clang-tools-extra/docs/ReleaseNotes.rst
+++ b/clang-tools-extra/docs/ReleaseNotes.rst
@@ -269,6 +269,10 @@ Changes in existing checks
   <clang-tidy/checks/modernize/loop-convert>` to support for-loops with
   iterators initialized by free functions like ``begin``, ``end``, or ``size``.
 
+- Improved :doc:`modernize-loop-convert
+   <clang-tidy/checks/modernize/loop-convert>` to avoid crash dor dependent
+   array.
+
 - Improved :doc:`modernize-return-braced-init-list
   <clang-tidy/checks/modernize/return-braced-init-list>` check to ignore
   false-positives when constructing the container with ``count`` copies of

--- a/clang-tools-extra/test/clang-tidy/checkers/modernize/loop-convert-basic.cpp
+++ b/clang-tools-extra/test/clang-tidy/checkers/modernize/loop-convert-basic.cpp
@@ -939,15 +939,18 @@ void fundamentalTypesTest() {
   // CHECK-FIXES: for (double Double : Doubles)
 }
 
-template <unsigned int p> void test() {
-  unsigned int test[3][p];
-  // Initialize to zero
-  for (unsigned int i = 0; i < p; ++i)
-    for (unsigned int j = 0; j < 3; ++j)
-      test[j][i] = 0;
+template <unsigned  p> void _dependenceArrayTest() {
+  unsigned test[3][p];
+  for (unsigned i = 0; i < p; ++i)
+    for (unsigned j = 0; j < 3; ++j)
+      printf("%d", test[j][i]);
   // CHECK-MESSAGES: :[[@LINE-2]]:5: warning: use range-based for loop instead
   // CHECK-FIXES: (auto & j : test)
-  // CHECK-FIXES: j[i] = 0;
+  // CHECK-FIXES: printf("%d", j[i]);
+}
+void dependenceArrayTest() {
+  _dependenceArrayTest<1>();
+  _dependenceArrayTest<2>();
 }
 
 } // namespace PseudoArray

--- a/clang-tools-extra/test/clang-tidy/checkers/modernize/loop-convert-basic.cpp
+++ b/clang-tools-extra/test/clang-tidy/checkers/modernize/loop-convert-basic.cpp
@@ -939,4 +939,13 @@ void fundamentalTypesTest() {
   // CHECK-FIXES: for (double Double : Doubles)
 }
 
+template <unsigned int p> void test() {
+  unsigned int test[3][p];
+  // Initialize to zero
+  for (unsigned int i = 0; i < p; ++i)
+    for (unsigned int j = 0; j < 3; ++j) // CHECK-MESSAGES: warning: use range-based for loop instead
+  // CHECK-FIXES: (auto & j : test)
+      test[j][i] = 0;
+}
+
 } // namespace PseudoArray

--- a/clang-tools-extra/test/clang-tidy/checkers/modernize/loop-convert-basic.cpp
+++ b/clang-tools-extra/test/clang-tidy/checkers/modernize/loop-convert-basic.cpp
@@ -943,9 +943,11 @@ template <unsigned int p> void test() {
   unsigned int test[3][p];
   // Initialize to zero
   for (unsigned int i = 0; i < p; ++i)
-    for (unsigned int j = 0; j < 3; ++j) // CHECK-MESSAGES: warning: use range-based for loop instead
-  // CHECK-FIXES: (auto & j : test)
+    for (unsigned int j = 0; j < 3; ++j)
       test[j][i] = 0;
+  // CHECK-MESSAGES: :[[@LINE-2]]:5: warning: use range-based for loop instead
+  // CHECK-FIXES: (auto & j : test)
+  // CHECK-FIXES: j[i] = 0;
 }
 
 } // namespace PseudoArray

